lib/modules.nix

```diff
@@ -424,12 +435,66 @@ rec {
         else if tp.name == "list of option sets" then types.listOf (types.submodule options)
         else if tp.name == "null or option set" then types.nullOr (types.submodule options)
         else tp;
+
     in
       if opt.type.getSubModules or null == null
-      then opt // { type = f (opt.type or types.unspecified); }
+      then opt // { type = f (opt.type or (inferType typeInference loc opt)); }
```

Comment:
I would prefer the type enforcement to be mandatory, we a clear release note message to explain the steps to NixOS developer on how they can change their modules to use `types.unspecified`.

As this is a non-backward compatible change, this should probably be a warning first in 16.03, and enforced in the next version of NixOS.  Having extra types is backward compatible, so I do not see the need for `_module.typeInference`.